EIPS/eip-1967.md

```diff
@@ -51,18 +57,35 @@ Holds the address that is allowed to upgrade the logic contract address for this
 
-Note that a clash may occur even among functions with different names, since the ABI relies on just four bytes for the function selector. This can lead to unexpected errors, or even exploits, where a call to a proxied contract returns a different value than expected, since the proxy intercepts the call and answers with a value of its own. 
+Note that a clash may occur even among functions with different names, since the ABI relies on just four bytes for the function selector. This can lead to unexpected errors, or even exploits, where a call to a proxied contract returns a different value than expected, since the proxy intercepts the call and answers with a value of its own.
 
 From [_Malicious backdoors in Ethereum proxies_](https://medium.com/nomic-labs-blog/malicious-backdoors-in-ethereum-proxies-62629adf3357) by Nomic Labs:
 
 > Any function in the Proxy contract whose selector matches with one in the implementation contract will be called directly, completely skipping the implementation code.
-> 
+>
 > Because the function selectors use a fixed amount of bytes, there will always be the possibility of a clash. This isn’t an issue for day to day development, given that the Solidity compiler will detect a selector clash within a contract, but this becomes exploitable when selectors are used for cross-contract interaction. Clashes can be abused to create a seemingly well-behaved contract that’s actually concealing a backdoor.
 
 The fact that proxy public functions are potentially exploitable makes it necessary to standardise the logic contract address in a different way. This approach is also used as part of [Universal Upgradeable Proxy Standard (EIP1822)](./eip-1822.md), which could become a specialization of this EIP.
 
+## Beacons
+
+Some use-cases rely on multiple proxy contracts delegating their calls to the same logic contract. If each contract was to store the address of the logic contract using the logic contract storage slot, upgrading them would require writing to the corresponding storage slot of every instance, which would be very inefficient.
+
+Another approach is to have each proxy retrieve the logic contract's address for a dedicated beacon. Using this pattern, the address of the logic contract can be modified in the beacon and take immediate effect for all the corresponding proxy contracts.
+
+A beacon MUST implements the function:
+
+```
+function implementation() returns (address)
+```
+
+Beacon based proxy contracts do not use the logic contract slot. Instead, they use the beacon contract slot to store the address of the beacon they are attached to. In order to know the logic contract used by a beacon proxy, one should:
+
+- Read the address of the beacon for the beacon logic storage slot;
+- Call the `implementation()` function on the beacon contract (with `msg.sender` set to be the proxy's address).
```

Comment:
Have you seen `msg.sender` used in a beacon?

I don't think this should be part of the specification. Off-chain you can simulate the sender using `eth_call`, but on-chain other contracts would not be able to retrieve the implementation if `msg.sender` changed the result. If anything I would specify that `implementation()` "MUST NOT" depend on `msg.sender`, or  "SHOULD NOT" if we want to allow people to do it if they weigh the consequences.